### PR TITLE
Publish to nexus and add provenance record

### PR DESCRIPTION
**Pull request by joeyzhou98**
Extend publishing functions to allow publishing to Nexus.
See #521

### CODE REVIEW

**Comment by joeyzhou98** [comment]
Hi @glatard, Nexus is organized in such a way so that there are organizations that can contain multiple projects, and in turn, those projects can contain multiple resources that have types like resolvers, schemas, views or files. This is described here: https://bluebrainnexus.io/docs/api/1.0/index.html. I have a few questions in term of the logistics of publishing to our Nexus server:
1) Should the deciding factors of publishing to which Nexus organizations and projects be manual, automatic or hard-coded? Manual would imply like specifying through the command line which organization and project to publish to, for example `bosh data publish --nexus --organization boutiques --project SIMON-dataset`. Automatic would imply reading the dataset/file title and creating/updating projects with that title so that you won't need to specify the project/organization, but I will need to know where to read the organization and project names from. Hard-coded would mean publishing to a specific organization and project every time.
2) Under what type of resources, mentioned above, should the files uploaded be a type of? Again we could make it so the code automatically decides based on some property of item in the files and I would need to know what are those properties, but again, specifying manually through flags is also an option

**Comment by glatard** [comment]
Hi @joeyzhou98, sorry for the delay responding to this.
1. I think we should treat organization and projects as publication parameters, similar to `zenodo-access-token`. So it should be possible to store them in a config file or to pass them on the command line. The command line should have precedence over the config file (if no param is passed on the command line then read config file). You should reuse the same logic as the one used for `zenodo-access-token`.
2. Data sounds like the right type to use.

File: tools/python/boutiques/bosh.py [hunk 1]
```diff
@@ -125,6 +126,9 @@ def execute(*params):
         parser.add_argument("--skip-data-collection", action="store_true",
                             help="Skips execution data collection and saving"
                             "to cache.")
+        parser.add_argument("--provenance", action="store", type=json.loads,
+                            help="Append provenance object to the "
```
Comment (glatard) [inline tools/python/boutiques/bosh.py]: Should be `Append JSON object to the generated record.`

File: tools/python/boutiques/nexusHelper.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        json_creds[self.config_token_property_name()] = access_token
+        with open(self.config_file, 'w') as f:
+            f.write(json.dumps(json_creds, indent=4, sort_keys=True))
+        if self.verbose:
+            print_info("Nexus access token saved in {0}".
+                       format(self.config_file))
+
+    def get_nexus_endpoint(self):
+        # change once nexus is setup
+        endpoint = "https://sandbox.bluebrainnexus.io/v1" if self.sandbox \
```
Comment (glatard) [inline tools/python/boutiques/nexusHelper.py]: Could you create an issue for that?

**Comment by joeyzhou98** [comment]
Hi, @glatard this is a big PR, it might take some time to review everything

File: tools/python/boutiques/nexusHelper.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        json_creds["nexus-project"] = project
+        with open(self.config_file, 'w') as f:
+            f.write(json.dumps(json_creds, indent=4, sort_keys=True))
+        if self.verbose:
+            print_info("Nexus access token, organization and project"
+                       " saved in {0}".format(self.config_file))
+
+    def get_nexus_endpoint(self):
+        # change once nexus is setup
+        endpoint = "https://sandbox.bluebrainnexus.io/v1" if self.sandbox \
```
Comment (glatard) [inline tools/python/boutiques/nexusHelper.py]: Could you mention this URL in the helptext so that users know where to create an account?

**Pull request merged**